pkg/provisioner/ironic/ironic.go

```diff
@@ -456,6 +456,20 @@ func (p *ironicProvisioner) ValidateManagementAccess(data provisioner.Management
 		// below.
 	}
 
+	// If no PreprovisioningImage builder is enabled we set the Node network_data
+	// this enables Ironic to inject the network_data into the ramdisk image
+	if !p.config.havePreprovImgBuilder {
+		p.log.Info("adding preprovisioning network_data for node", "node", ironicNode.UUID)
+		networkDataRaw := data.PreprovisioningNetworkData
+		if networkDataRaw != "" {
+			var networkData map[string]interface{}
+			if err := yaml.Unmarshal([]byte(networkDataRaw), &networkData); err != nil {
+				p.log.Error(err, "failed to unmarshal network_data.json from PreprovisioningNetworkData")
```

Comment:
Should we return `operationFailed` here or `transientError` since the secret contents may get updated?

Reply:
That's not the topic I'm strong at... I guess we'll get another reconcile if it gets updated (assuming we listen to these events).

Reply:
Yeah that's what I wasn't sure of, I know we'll get another reconcile if the BMH changes but I wasn't sure if it's only the content of a referenced secret - I'll test to confirm

Reply:
We'll get another reconcile if the Secret changes because of https://github.com/metal3-io/baremetal-operator/blob/main/controllers/metal3.io/baremetalhost_controller.go#L1741, provided that we own the Secret, which is set up by https://github.com/metal3-io/baremetal-operator/blob/main/controllers/metal3.io/baremetalhost_controller.go#L648 - and note that this is currently only called for the preprovisioningNetworkData after we see it show up in the PreprovisioningImage Status. So we if want to use it all the time then we should call AcquireSecret somewhere unconditional.

That said, we don't have great ways of handling errors in general when the error is user configuration. The validating webhook is the best place to catch those. Also, this should be logged at info, not error level.
What would we be hoping to achieve here by returning early? Will ironic reject the update if the data is not yaml? If not I'd advocate that we continue to do what the patch currently does - pass whatever garbage the user gave us on to ironic and carry on.

Reply:
> Will ironic reject the update if the data is not yaml? 

It [validates it against the network_data.json schema](https://opendev.org/openstack/ironic/src/branch/master/ironic/api/controllers/v1/node.py#L394) so the format has to be exactly right or we'll get an error on node registration from Ironic

I discovered this while testing so decided to return early here in an attempt to catch obviously incorrect secret contents, it also matches what we do on [deployment for the config-drive network_data](https://github.com/metal3-io/baremetal-operator/blob/main/pkg/provisioner/ironic/ironic.go#L1477)

Reply:
OK, we're returning transientError in the config-drive case, so there's precedent for that even if it is not really the right thing.

Reply:
Ok I went with `operationFailed` in the latest revision but will adjust to `transientError` since it sounds like that will work without adjusting the secret ownership.

Related to that, I thought of a disadvantage of adjusting the network data ownership - that will result in behaviour similar to https://github.com/metal3-io/baremetal-operator/issues/1347 right?  Which might be desirable in the static-ip case (one secret per BMH) but probably not in the case where there's a config applied per cluster or group of hosts (non-default vlan, bonding etc)

Reply:
> Ok I went with `operationFailed` in the latest revision but will adjust to `transientError` since it sounds like that will work without adjusting the secret ownership.

You'll want to adjust it anyway for the cases where there is _not_ an error.

> Related to that, I thought of a disadvantage of adjusting the network data ownership - that will result in behaviour similar to #1347 right?

Yes.

> Which might be desirable in the static-ip case (one secret per BMH) but probably not in the case where there's a config applied per cluster or group of hosts (non-default vlan, bonding etc)

If there's multiple owners, the Secret doesn't get deleted until the last owner is. I think it's fine. This is consistent with how e.g. networkData works, and with how preprovisioningNetworkData works when the PreprovisioningImage controller is enabled.